packages/expo-doctor/src/utils/explainDependencies.ts

```diff
@@ -0,0 +1,132 @@
+import spawnAsync, { SpawnResult } from '@expo/spawn-async';
+import chalk from 'chalk';
+import semver from 'semver';
+
+import { RootNodePackage, VersionSpec } from './explainDependencies.types';
+
+type TargetPackage = { name: string; version?: VersionSpec };
+
+function isSpawnResult(result: any): result is SpawnResult {
+  return 'stderr' in result && 'stdout' in result && 'status' in result;
+}
+
+/** Spawn `npm explain [name] --json` and return the parsed JSON. Returns `null` if the requested package is not installed. */
+async function explainAsync(
+  packageName: string,
+  projectRoot: string,
+  parameters: string[] = []
+): Promise<RootNodePackage[] | null> {
+  const args = ['explain', packageName, ...parameters, '--json'];
+
+  try {
+    const { stdout } = await spawnAsync('npm', args, {
+      stdio: 'pipe',
+      cwd: projectRoot,
+    });
+
+    return JSON.parse(stdout);
+  } catch (error: any) {
+    if (isSpawnResult(error)) {
+      if (error.stderr.match(/npm ERR! No dependencies found matching/)) {
+        return null;
+      } else if (error.stdout.match(/Usage: npm <command>/)) {
+        throw new Error(
+          `Dependency tree validation for ${chalk.underline(
+            packageName
+          )} failed. This validation is only available on Node 16+ / npm 8.`
+        );
+      }
+    }
+    throw new Error(`Failed to find dependency tree for ${packageName}: ` + error.message);
+  }
+}
+
+function organizeExplanations(
+  pkg: TargetPackage,
+  {
+    explanations,
+    isValid,
+  }: {
+    explanations: RootNodePackage[];
+    isValid: (pkg: TargetPackage) => boolean;
+  }
+) {
+  const valid: RootNodePackage[] = [];
+  const invalid: RootNodePackage[] = [];
+
+  for (const explanation of explanations) {
+    const { name } = explanation;
+    if (name === pkg.name) {
+      if (isValid(explanation)) {
+        valid.push(explanation);
+      } else {
+        invalid.push(explanation);
+      }
+    }
+  }
+  return { valid, invalid };
+}
+
+/**
+ * @param pkg
+ * @param explanations
+ * @returns true if all versions of the package satisfy the constraints
+ */
+async function printExplanationsAsync(
+  pkg: TargetPackage,
+  explanations: RootNodePackage[]
+): Promise<string | null> {
+  const { invalid } = organizeExplanations(pkg, {
+    explanations,
+    isValid(otherPkg) {
+      return semver.satisfies(otherPkg.version!, pkg.version!);
+    },
+  });
+
+  if (invalid.length > 0) {
+    return getInvalidPackagesWarning(pkg, { explanations: invalid });
+  }
+  return null;
+}
+
+function getInvalidPackagesWarning(
+  pkg: TargetPackage,
+  { explanations }: { explanations: RootNodePackage[] }
+): string {
+  const lines = [];
+  if (pkg.version) {
+    lines.push(`Expected package ${formatPkg(pkg, 'green')}`);
+  } else {
+    lines.push(`Expected to not find any copies of ${formatPkg(pkg, 'green')}`);
+  }
```

Comment:
This logic is very hard to follow. It depends on a calling convention of the `warnAbout...` method several places up in the stack, that will only include a version for a package if the versions endpoint has that package listed. And even that logic only 'works' for the specific packages that we use the `warnAbout...` method with (illegalPackages - always without a version number, and supportPackages - only with a version number if they were still in use in a particular SDK version).

I think it would be much clearer to explicitly establish the results you are expecting in this call chain. You're looking for a package, and you either don't want to find it, or you want to find a specific version of it. This expectation currently only becomes clear here, 3 levels deep, after passing down through `warnAboutDeepDependenciesAsync` and `printExplanationsAsync`.